File: play-2.5.x/K-Modules/kmodules/K-Base/app/models/K/EbeanConfig/MyBeanPersistController.java
package models.K.EbeanConfig;

import K.Aop.annotations.WithPersistLog;
import K.Common.Helper;
import com.avaje.ebean.event.BeanPersistController;
import com.avaje.ebean.event.BeanPersistRequest;
import com.google.common.collect.ImmutableSet;
import com.google.common.reflect.ClassPath;
import play.Logger;
import play.libs.Json;

import javax.persistence.Entity;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by kk on 14/11/28.
 */
public class MyBeanPersistController implements BeanPersistController {

    private Set<String> registed_bean_class_names;

    public MyBeanPersistController() {
        Logger.debug("Ebean server load MyBeanPersistController");
        registed_bean_class_names = new HashSet<>();
        try {
            ClassPath cp = ClassPath.from(MyBeanPersistController.class.getClassLoader());
            String ModelsPackage = "models";
            ImmutableSet<ClassPath.ClassInfo> classes = cp.getTopLevelClassesRecursive(ModelsPackage);
            for (ClassPath.ClassInfo c : classes) {
                Class<?> BeanClass = c.load();
                Entity entityAnno = BeanClass.getAnnotation(Entity.class);
                if (entityAnno == null) {
                    continue;
                }
                WithPersistLog withPersistLogAnno = BeanClass.getAnnotation(WithPersistLog.class);
                if (withPersistLogAnno == null) {
                    continue;
                }
                registed_bean_class_names.add(BeanClass.getName());
//                Logger.debug("MyBeanPersistController regist bean class: {}", BeanClass.getName());
            }
        } catch (Exception ex) {
            Logger.error(ex.getMessage());
        }
    }

    // When there are multiple BeanPersistController's for a given entity type this controls the order in which they are executed.
    // Lowest values are executed first.
    @Override
    public int getExecutionOrder() {
        return 10;
    }

    @Override
    public boolean isRegisterFor(Class<?> cls) {
        return registed_bean_class_names.contains(cls.getName());
    }

    @Override
    public boolean preInsert(BeanPersistRequest<?> request) {
        return true;
    }

    @Override
    public boolean preUpdate(BeanPersistRequest<?> request) {
        return true;
    }

    @Override
    public boolean preDelete(BeanPersistRequest<?> request) {
        return true;
    }

    @Override
    public void postInsert(BeanPersistRequest<?> request) {
        WriteLog(request, request.getBean(), "Insert");
    }

    @Override
    public void postUpdate(BeanPersistRequest<?> request) {
        WriteLog(request, request.getBean(), "Update");
    }

    @Override
    public void postDelete(BeanPersistRequest<?> request) {
        WriteLog(request, request.getBean(), "Delete");
    }

    private void WriteLog(BeanPersistRequest<?> request, Object bean, String action_type) {
        PersistLog plog = new PersistLog();
        plog.bean_class = bean.getClass().getName();
        plog.action_type = action_type;
        plog.bean_data = Json.toJson(bean);
        StringBuilder sb = getBuffer(request);
        sb.append(Helper.ToJsonStringPretty(plog)).append("\n");

    }

    private StringBuilder getBuffer(BeanPersistRequest<?> request) {
        Object obj = request.getTransaction().getUserObject(MyBeanPersistController.class.getName());
        if (obj == null) {
            StringBuilder sb = new StringBuilder();
            request.getTransaction().putUserObject(MyBeanPersistController.class.getName(), sb);
            return sb;
        } else {
            return (StringBuilder) obj;
        }
    }
}

File: play-kotlin-2.5.x/PlayKotlin/kmodules/K-Base/app/models/K/EbeanConfig/MyBeanPersistController.kt
package models.K.EbeanConfig

import K.Aop.annotations.WithPersistLog
import K.Common.Helper
import com.avaje.ebean.event.BeanPersistController
import com.avaje.ebean.event.BeanPersistRequest
import com.google.common.collect.ImmutableSet
import com.google.common.reflect.ClassPath
import play.Logger
import play.libs.Json

import javax.persistence.Entity
import java.util.HashSet

/**
 * Created by kk on 14/11/28.
 */
class MyBeanPersistController : BeanPersistController {

    private val registed_bean_class_names: MutableSet<String>

    init {
        Helper.DLog("Ebean server load MyBeanPersistController")
        registed_bean_class_names = HashSet<String>()
        try {
            val cp = ClassPath.from(MyBeanPersistController::class.java.classLoader)
            val ModelsPackage = "models"
            val classes = cp.getTopLevelClassesRecursive(ModelsPackage)
            for (c in classes) {
                val BeanClass = c.load()
                val entityAnno = BeanClass.getAnnotation(Entity::class.java) ?: continue
                val withPersistLogAnno = BeanClass.getAnnotation(WithPersistLog::class.java) ?: continue
                registed_bean_class_names.add(BeanClass.name)
                //                Logger.debug("MyBeanPersistController regist bean class: {}", BeanClass.getName());
            }
        } catch (ex: Exception) {
            Logger.error(ex.message)
        }

    }

    // When there are multiple BeanPersistController's for a given entity type this controls the order in which they are executed.
    // Lowest values are executed first.
    override fun getExecutionOrder(): Int {
        return 10
    }

    override fun isRegisterFor(cls: Class<*>): Boolean {
        return registed_bean_class_names.contains(cls.name)
    }

    override fun preInsert(request: BeanPersistRequest<*>): Boolean {
        return true
    }

    override fun preUpdate(request: BeanPersistRequest<*>): Boolean {
        return true
    }

    override fun preDelete(request: BeanPersistRequest<*>): Boolean {
        return true
    }

    override fun postInsert(request: BeanPersistRequest<*>) {
        WriteLog(request, request.bean, "Insert")
    }

    override fun postUpdate(request: BeanPersistRequest<*>) {
        WriteLog(request, request.bean, "Update")
    }

    override fun postDelete(request: BeanPersistRequest<*>) {
        WriteLog(request, request.bean, "Delete")
    }


    private fun WriteLog(request: BeanPersistRequest<*>, bean: Any, action_type: String) {
        val plog = PersistLog()
        plog.bean_class = bean.javaClass.name
        plog.action_type = action_type
        plog.bean_data = Json.toJson(bean)
        val sb = getBuffer(request)
        sb.append(Helper.ToJsonStringPretty(plog)).append("\n")

    }

    private fun getBuffer(request: BeanPersistRequest<*>): StringBuilder {
        val obj = request.transaction.getUserObject(MyBeanPersistController::class.java.name)
        if (obj == null) {
            val sb = StringBuilder()
            request.transaction.putUserObject(MyBeanPersistController::class.java.name, sb)
            return sb
        } else {
            return obj as StringBuilder
        }
    }
}
